src/appleseed-max-impl/appleseedoslplugin/oslparamdlg.cpp

```diff
@@ -0,0 +1,260 @@
+#include "utilities.h"
+
+// 3ds Max headers.
+#include <iparamm2.h>
+#include <maxtypes.h>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+OSLParamDlg::OSLParamDlg(
+    HWND                    hwMtlEdit, 
+    IMtlParams*             imp, 
+    MtlBase*                map,
+    const OSLShaderInfo*    shader_info)
+  : m_hmedit(hwMtlEdit)
+  , m_bump_pmap(nullptr)
+  , m_pmap(nullptr)
+  , m_imp(imp)
+  , m_osl_plugin(map)
+  , m_shader_info(shader_info)
+  , m_col_1(10)
+  , m_col_2(85)
+  , m_col_3(122)
+  , m_col_4(159)
+  , m_label_width(76)
+  , m_edit_width(25)
+  , m_edit_height(10)
+  , m_tex_width(84)
+{
+    m_class_id = m_osl_plugin->ClassID();
+    create_dialog();
+}
+
+Class_ID OSLParamDlg::ClassID()
+{
+    return m_class_id;
+}
+
+ReferenceTarget* OSLParamDlg::GetThing()
+{
+    return m_osl_plugin;
+}
+
+void OSLParamDlg::SetThing(ReferenceTarget *m)
+{
+    assert(m->ClassID() == m_osl_plugin->ClassID());
+    m_osl_plugin = static_cast<MtlBase*>(m);
+    m_pmap->SetParamBlock(m_osl_plugin->GetParamBlock(0));
+}
+
+void OSLParamDlg::DeleteThis()
+{
+    if (m_pmap != nullptr)
+        DestroyMParamMap2(m_pmap);
+    
+    if (m_bump_pmap != nullptr)
+        DestroyMParamMap2(m_bump_pmap);
+
+    m_pmap = nullptr;
+    m_bump_pmap = nullptr;
+    delete this;
+}
+
+void OSLParamDlg::SetTime(TimeValue t) {}
+void OSLParamDlg::ReloadDialog(void) {}
+void OSLParamDlg::ActivateDlg(BOOL onOff) {}
+
+void OSLParamDlg::add_groupboxes(
+    DialogTemplate&         dialog_template,
+    PageGroupMap&           pages)
+{
+    int y_pos = 17;
+    for (auto& page : pages)
+    {
+        int height = page.second.height + 10;
+        page.second.y_pos = y_pos + 10;
+        dialog_template.AddButton((LPCSTR)page.first.c_str(), WS_VISIBLE | BS_GROUPBOX | WS_CHILD | WS_GROUP, NULL, 4, y_pos, 209, height, 0);
+        y_pos += height;
+        y_pos += 4;
+    }
+}
+
+void OSLParamDlg::add_ui_parameter(
+    DialogTemplate&         dialog_template,
+    const MaxParam&         max_param,
+    PageGroupMap&           pages)
+{
+    int y_pos = 5;
+    auto param_page = pages.find(max_param.page_name);
+    if (param_page != pages.end())
+        y_pos = param_page->second.y_pos;
+
+    int ctrl_id = max_param.max_ctrl_id;
+    const std::string param_label = max_param.max_label_str + ":";
+    dialog_template.AddStatic((LPCSTR)param_label.c_str(), WS_VISIBLE, NULL, m_col_1, y_pos, m_label_width, m_edit_height, ctrl_id++);
+    switch (max_param.param_type)
+    {
+      case MaxParam::Float:
+      case MaxParam::IntNumber:
+        dialog_template.AddComponent((LPCSTR)"CustEdit", (LPCSTR)"Edit", WS_VISIBLE | WS_TABSTOP, NULL, m_col_2, y_pos, m_edit_width, m_edit_height, ctrl_id++);
+        dialog_template.AddComponent((LPCSTR)"SpinnerControl", (LPCSTR)"Spinner", WS_VISIBLE | WS_TABSTOP, NULL, m_col_2 + m_edit_width + 2, y_pos, 7, m_edit_height, ctrl_id++);
+        break;
+      case MaxParam::VectorParam:
```

Comment:
I suggest inserting one blank line between the various cases (our case blocks) to make this code easier to read.